For #800: Fix FtRemote for empty response body test case and unignore returnsAnEmptyResponseBody.

paulodamaso opened the pull request:
For #800:
- Fixed test cases for empty response body;
- Removed `@todo` and `@Ignore` from tests

0crat commented:
Job [#855](https://github.com/yegor256/takes/issues/855) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2641744261, version: 0.33.4, hash: ${buildNumber} -->

0crat commented:
This pull request [#855](https://github.com/yegor256/takes/issues/855) is assigned to @g4s8[/z](https://www.0crat.com/u/g4s8), here is [why](https://www.0crat.com/footprint/C3FFK3YAY/2641756440); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2641950402, version: 0.33.4, hash: ${buildNumber} -->

g4s8 commented on src/test/java/org/takes/http/FtRemoteTest.java:
```diff
@@ -148,7 +142,7 @@ public void exec(final URI home) throws IOException {
                         .fetch()
                         .as(RestResponse.class)
-                        .assertBody(new IsEqual<>(""))
+                        .assertBody(new IsEqual<>("HTTP/1.1 200 OK\r\n\r\n"))
```
@paulodamaso it looks like a head of response, not body. According to test method name the response body should be empty.

paulodamaso replied:
@g4s8 AssertBody method does not assert just the body of the response. `RsPrint` merges head and body and for that moment on it is treated like a single thing.  So I'm testing if we have just the headers in this response.

paulodamaso replied:
@g4s8 AssertBody method does not assert just the body of the response. RsPrint merges head and body and for that moment on it is treated like a single thing. So I'm testing if we have just the headers in this response.

t-izbassar replied:
@paulodamaso I agree with @g4s8. The test now is misleading as it says that `TkEmpty` should return empty body in response, but that's not what you assertion is proven. This assertion proves that we have some bug (which is the fact that response body is not empty even though it should be empty).

g4s8 commented on src/test/java/org/takes/tk/TkSlf4jRemoteTest.java:
```diff
@@ -62,7 +61,7 @@ public void exec(final URI home) throws IOException {
                         .fetch()
                         .as(RestResponse.class)
-                        .assertBody(new IsEqual<>(""))
+                        .assertBody(new IsEqual<>("HTTP/1.1 200 OK\r\n\r\n"))
```
@paulodamaso same here

paulodamaso commented:
@g4s8 answered, please take a look

t-izbassar commented on pom.xml:
```diff
@@ -254,7 +254,7 @@ SOFTWARE.
         <artifactId>maven-surefire-plugin</artifactId>
         <configuration>
-          <argLine>${argLine} -Djava.awt.headless=true -Dfile.encoding=UTF-8</argLine>
+          <argLine>${argLine} -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Duser.language=en -Duser.region=US</argLine>
```
I don't think that hard-coding language and region is a good idea. Our tests should not depend on that.

paulodamaso replied:
@t-izbassar On the contrary, putting this encoding and region info ensures that our tests work as intended. We are heaving a lot of problems in other projects due to language issues (mostly number formatting, some locales use "," as decimal point, some use ".") and these setting these locale info on the project assures that we will always use the correct one despite the user machine configuration.

t-izbassar replied:
@paulodamaso I think the tests that require the language should set it explicitly in tests so that they are visible. But putting global hard-coded value is a mistake.

paulodamaso commented:
@g4s8 @t-izbassar As asked, I've reverted the changes and added an issue to jcabi-http project, and changed our issue so it can only be fixed after we deal with the problems in `assertBody` in `jcabi-http`. I've reverted `pom.xml` changes because although I think that they're pertinent and important they does not belong to this ticket. Please take a look now and say me what you think.

t-izbassar commented on src/test/java/org/takes/http/FtRemoteTest.java:
```diff
@@ -52,10 +52,11 @@
+ *  some bug in jcabi-http (reported in
+ *  https://github.com/jcabi/jcabi-http/issues/177). After this bug is fixed,
+ *  correct the testing for this condition in returnsAnEmptyResponseBody and
+ *  TkSlf4jRemoteTest.
```
@paulodamaso Please preserve the
> After that the
 *  {@link org.takes.tk.TkSlf4jRemoteTest} should be fixed and relevant
 *  test case should be unignored.

from original puzzle as otherwise the original test case will be forgotten. Also I don't think that bug in `jcabi-http` is what we have. Have you tried it with different http client? Also does it will fix the original issue #712?

paulodamaso commented:
@t-izbassar I don't know if it fixes https://github.com/yegor256/takes/issues/712, I'm trying to  find why https://github.com/yegor256/takes/issues/800 is happening and these classes are not behaving accordingly.

I've not tried with another http client either, but I've tracked the problem to the `jcabi-http` library and documented it in ticket comments and in PR comments too. I've spent way more than 30 minutes in this ticket, mostly in review, and this issue is clogging my agenda for almost a month now.

If the presented solutions are not enough, please tell me precisely what would I do so we could close this ticket. Takes tickets are stale in my agenda and are preventing me for making money, so please help me to get rid of this one.

0crat commented:
The architect of the project has changed; @yegor256[/z](https://www.0crat.com/u/yegor256) is not at this role anymore; @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) is the architect now

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2831263163, version: 0.35.5, hash: ${buildNumber} -->

paulodamaso commented:
@g4s8 @t-izbassar ping, what we need to get this one moving?

t-izbassar commented:
@paulodamaso as long as you are an architect now - it’s your call to make. I’ve already expressed my view. I don’t think that issue with `jcabi-http` is connected to the original problem. At least there is no test added that proves it. You haven’t checked if problem remains with different http client so from my perspective - this PR do not solve anything.

0crat commented:
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @g4s8[/z](https://www.0crat.com/u/g4s8), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2860117396, version: 0.35.5, hash: ${buildNumber} -->

0crat commented:
Code review was too long (21 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2860117402, version: 0.35.5, hash: ${buildNumber} -->

0crat commented:
The job [#855](https://github.com/yegor256/takes/issues/855) is now out of scope

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2860117411, version: 0.35.5, hash: ${buildNumber} -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

<!-- https://www.0crat.com/footprint/C3FFK3YAY/2860117426, version: 0.35.5, hash: ${buildNumber} -->

elenavolokhova commented:
@g4s8 According to our Policy:
> The code reviewer found at least three problems in the code 

Only two issues were found during this review.
Please, confirm that you will try to find at least three issues next time.

elenavolokhova commented:
@0crat quality acceptable
